src/VisualStudio/Core/Def/TaskList/ExternalErrorDiagnosticUpdateSource.cs

```diff
@@ -5,132 +5,103 @@
         IThreadingContext threadingContext)
-        : this(workspace, diagnosticService, notificationService, listenerProvider.GetListener(FeatureAttribute.ErrorList), threadingContext.DisposalToken)
-    {
-    }
-
-    /// <summary>
-    /// internal for testing
-    /// </summary>
-    internal ExternalErrorDiagnosticUpdateSource(
-        Workspace workspace,
-        IDiagnosticAnalyzerService diagnosticService,
-        IGlobalOperationNotificationService notificationService,
-        IAsynchronousOperationListener listener,
-        CancellationToken disposalToken)
     {
-        // use queue to serialize work. no lock needed
-        _taskQueue = new TaskQueue(listener, TaskScheduler.Default);
-        _disposalToken = disposalToken;
-
+        _disposalToken = threadingContext.DisposalToken;
         _workspace = workspace;
-        _workspace.WorkspaceChanged += OnWorkspaceChanged;
-
         _diagnosticService = diagnosticService;
-        _buildOnlyDiagnosticsService = _workspace.Services.GetRequiredService<IBuildOnlyDiagnosticsService>();
+        _listener = listenerProvider.GetListener(FeatureAttribute.ErrorList);
+
+        _serviceBroker = serviceBroker;
+        _taskQueue = new AsyncBatchingWorkQueue<Func<CancellationToken, Task>>(
+            TimeSpan.Zero,
+            processBatchAsync: ProcessTaskQueueItemsAsync,
+            _listener,
+            _disposalToken
+        );
+    }
 
-        _notificationService = notificationService;
+    private async ValueTask ProcessTaskQueueItemsAsync(ImmutableSegmentedList<Func<CancellationToken, Task>> list, CancellationToken cancellationToken)
+    {
+        foreach (var workItem in list)
+            await workItem(cancellationToken).ConfigureAwait(false);
     }
 
     public DiagnosticAnalyzerInfoCache AnalyzerInfoCache => _diagnosticService.AnalyzerInfoCache;
 
-    /// <summary>
-    /// Event generated from the serialized <see cref="_taskQueue"/> whenever the build progress in Visual Studio changes.
-    /// Events are guaranteed to be generated in a serial fashion, but may be invoked on any thread.
-    /// </summary>
-    public event EventHandler<BuildProgress>? BuildProgressChanged;
-
-    /// <summary>
-    /// Event generated from the serialized <see cref="_taskQueue"/> whenever build-only diagnostics are reported during a build in Visual Studio.
-    /// These diagnostics are not supported from intellisense and only get refreshed during actual build.
-    /// </summary>
-    public event EventHandler<ImmutableArray<DiagnosticsUpdatedArgs>>? DiagnosticsUpdated;
-
-    /// <summary>
-    /// Indicates if a build is currently in progress inside Visual Studio.
-    /// </summary>
-    public bool IsInProgress => GetBuildInProgressState() != null;
-
     public void Dispose()
     {
         lock (_gate)
         {
-            _activeCancellationSeriesDoNotAccessDirectly.Dispose();
+            _diagnosticManagerService?.Dispose();
```

Comment:
i'm not sure i really get thsi part.  our Dispose is called when mef is tearing everything down.  Why do we need to have the diag manager do anything at that point?

Reply:
Nit: i would be 100% ok not doing this and just keeping things simple.  does anything go wrong if you remove Dispose fro this type *entirely*?